system/system_monitor/src/net_monitor/net_monitor.cpp

```diff
@@ -372,5 +378,81 @@ void NetMonitor::monitorTraffic(diagnostic_updater::DiagnosticStatusWrapper & st
   SystemMonitorUtility::stopMeasurement(t_start, stat);
 }
 
+void NetMonitor::checkIpReasmFails(diagnostic_updater::DiagnosticStatusWrapper & stat)
+{
+  // Remember start time to measure elapsed time
+  const auto t_start = SystemMonitorUtility::startMeasurement();
+
+  int whole_level = DiagStatus::OK;
+  std::string error_str;
+  uint64_t total_reasm_fails = 0;
+  uint64_t unit_reasm_fails = 0;
+
+  if (getIpReasmFails(total_reasm_fails)) {
+    reasm_fails_queue_.push_back(total_reasm_fails - last_reasm_fails_);
+    while (reasm_fails_queue_.size() > reasm_fails_check_duration_) {
+      reasm_fails_queue_.pop_front();
+    }
+
+    for (auto reasm_fails : reasm_fails_queue_) {
+      unit_reasm_fails += reasm_fails;
+    }
+
+    stat.add(fmt::format("total packet reassembles failed"), total_reasm_fails);
+    stat.add(fmt::format("packet reassembles failed per unit time"), unit_reasm_fails);
+
+    if (unit_reasm_fails >= reasm_fails_count_threshold_) {
+      whole_level = std::max(whole_level, static_cast<int>(DiagStatus::WARN));
+      error_str = "reassembles failed";
+    }
+
+    last_reasm_fails_ = total_reasm_fails;
+  } else {
+    reasm_fails_queue_.push_back(0);
+    whole_level = std::max(whole_level, static_cast<int>(DiagStatus::ERROR));
+    error_str = "failed to read /proc/net/snmp";
+  }
+
+  if (!error_str.empty()) {
+    stat.summary(whole_level, error_str);
+  } else {
+    stat.summary(whole_level, "OK");
+  }
+
+  // Measure elapsed time since start time and report
+  SystemMonitorUtility::stopMeasurement(t_start, stat);
+}
+
+bool NetMonitor::getIpReasmFails(uint64_t & reasm_fails)
+{
+  std::ifstream ifs("/proc/net/snmp");
+  if (!ifs) {
+    return false;
+  }
+
+  std::string line;
+  while (std::getline(ifs, line)) {
```

Comment:
This is deep nesting. Please separate to get titles and get the number of `ReasmFails`.
[pseudo code]
1. Read 1 line to get titles.
2. Find index of `ReasmFails`.
    If no index returned, return false;
4. Begin loop.
5. When the 1t item is `Ip:`, get the value of `ReasmFails` by using index.

Reply:
@ito-san 
I improved deep nesting.
24a196af79dda9c44c2781252396b25fd5df32f2

Reply:
`std::getline` in `std::getline` may cause confusion, and only one statement should be placed in the processing scope to prevent confusion.
I recommend as follows;
```cpp
bool NetMonitor::getReassemblesFailed(uint64_t & reassembles_failed)
{
  std::ifstream ifs("/proc/net/snmp");
  if (!ifs) {
    return false;
  }

  // /proc/net/snmp
  // Ip: Forwarding DefaultTTL InReceives ... ReasmTimeout ReasmReqds ReasmOKs ReasmFails ...
  // Ip: 2          64         5636471397 ... 135          2303339    216166   270        ...
  std::string line;

  if (!std::getline(ifs, line)) {
    return false;
  }

  // Find index of 'ReasmFails'
  std::vector<std::string> title_list;
  boost::split(title_list, line, boost::is_space());

  if (title_list.empty()) {
    return false;
  }
  if (title_list[0] != "Ip:") {
    return false;
  }

  int index = 0;
  for (const auto itr = title_list.begin(); itr != title_list.end(); ++itr, ++index) {
    if (title_list[index] == "ReasmFails") {
      break;
    }
  }

  // Find a value of 'ReasmFails'
  std::vector<std::string> value_list;
  boost::split(value_list, line, boost::is_space());

  if (!std::getline(ifs, line)) {
    return false;
  }

  if (title_list.size() != value_list.size()) {
    return false;
  }

  reassembles_failed = std::stoull(value_list[index]);

  return true;
}
```
In addition, the index to `ReasmFails` should be obtained at the beginning, not at this callback. Keeping the callback more light weight would be better. Please consider to fix it.

Reply:
`std::getline` in `std::getline` loop may cause confusion, and only one statement should be placed in the processing scope.
I recomment as follows;
```cpp
bool NetMonitor::getReassemblesFailed(uint64_t & reassembles_failed)
{
  std::ifstream ifs("/proc/net/snmp");
  if (!ifs) {
    return false;
  }

  // /proc/net/snmp
  // Ip: Forwarding DefaultTTL InReceives ... ReasmTimeout ReasmReqds ReasmOKs ReasmFails ...
  // Ip: 2          64         5636471397 ... 135          2303339    216166   270        ...
  std::string line;

  // Find index of 'ReasmFails'
  if (!std::getline(ifs, line)) {
    return false;
  }

  std::vector<std::string> title_list;
  boost::split(title_list, line, boost::is_space());

  if (title_list.empty()) {
    return false;
  }
  if (title_list[0] != "Ip:") {
    return false;
  }

  int index = 0;
  for (auto itr = title_list.begin(); itr != title_list.end(); ++itr, ++index) {
    if (*itr == "ReasmFails") {
      break;
    }
  }
  // Find a value of 'ReasmFails'
  if (!std::getline(ifs, line)) {
    return false;
  }

  std::vector<std::string> value_list;
  boost::split(value_list, line, boost::is_space());

  if (title_list.size() != value_list.size()) {
    return false;
  }

  reassembles_failed = std::stoull(value_list[index]);

  return true;
}
```
In addition, the index to `ReasmFails` should be obtained at the beginning, not at this callback. Keeping this callback more light weight would be better. Please consider to fix it.

Reply:
@ito-san 
I replaced it with the code you provided.
2c33f22f1282476884173aa36baff81e710b9a79

Reply:
@v-nakayama7440-esol 
Please address this issue.
> the index to `ReasmFails` should be obtained at the beginning, not at this callback. Keeping this callback more light weight would be better.

This processing should be moved to the constructor and wrapped into a function.
This only needs to be executed once in the lifetime of node because the index does not change on every callback cycle.  
```cpp
  int index = 0;
  for (auto itr = title_list.begin(); itr != title_list.end(); ++itr, ++index) {
    if (*itr == "ReasmFails") {
      break;
    }
  }
```

Reply:
@ito-san 
I fixed `ReasonFails` index obtained in the constructor.
dcf8b6c41b1e48f3882e658e6457282a97e9e232